File: src/a3/descriptor.ts
import type { StepId } from "../domain/model";

/**
 * D-03/D-04: the one artifact that feeds both `HtmlA3Renderer` (React, CSS
 * Grid) and the Rust `xlsx` writer (a dumb serializer, zero layout logic).
 * Every field here must be plain, JSON-serializable data — no functions, no
 * `Date`, no `Map`/`Set` — because it crosses the Tauri IPC boundary as JSON.
 */

export type BorderWeight = "thin" | "medium" | "thick";

export interface CellBorder {
  readonly top?: BorderWeight;
  readonly bottom?: BorderWeight;
  readonly left?: BorderWeight;
  readonly right?: BorderWeight;
  /** ARGB hex, e.g. "FF000000". Defaults to black in the renderer/writer if omitted. */
  readonly color?: string;
}

export type HorizontalAlign = "left" | "center" | "right";
export type VerticalAlign = "top" | "center" | "bottom";

export interface CellFont {
  readonly name: string;
  readonly sizePt: number;
  readonly bold?: boolean;
  readonly italic?: boolean;
  /** ARGB hex, e.g. "FF000000". */
  readonly color?: string;
}

export interface CellStyle {
  readonly id: string;
  readonly font: CellFont;
  /** ARGB hex fill, e.g. "FFFF0000". Omitted = no fill. */
  readonly fillColor?: string;
  readonly horizontalAlign?: HorizontalAlign;
  readonly verticalAlign?: VerticalAlign;
  readonly wrapText?: boolean;
  readonly border?: CellBorder;
}

/** A1-style cell reference, e.g. "B8". Always within the sheet's column/row grid. */
export type CellRef = string;
/** A1-style range, e.g. "B8:O8". */
export type RangeRef = string;

export interface CellData {
  readonly ref: CellRef;
  readonly value: string | number | null;
  readonly styleId?: string;
}

export interface MergedRange {
  readonly range: RangeRef;
}

export interface ColumnDef {
  readonly key: string;
  /** Native Excel character-width unit — passed to rust_xlsxwriter verbatim, no conversion. */
  readonly charWidth: number;
}

export interface RowDef {
  readonly index: number;
  readonly heightPt: number;
}

export interface ImagePlacement {
  readonly id: string;
  /** Base64-encoded image bytes. The writer/renderer never reads from disk (D-04). */
  readonly data: string;
  /**
   * D-118/D-193: `"image/jpeg"` for an ingested photo embedded straight
   * from its already-stored bytes (never rasterized) — `rust_xlsxwriter`'s
   * `Image::new_from_buffer` auto-detects the format from the buffer's own
   * signature, so this field is metadata only, not consulted by the writer
   * (`src-tauri/src/xlsx/writer.rs`), same as before this addition.
   */
  readonly mimeType: "image/png" | "image/jpeg";
  readonly anchorCell: CellRef;
  readonly offsetXPt?: number;
  readonly offsetYPt?: number;
  readonly widthPt: number;
  readonly heightPt: number;
}

export interface PageMarginsIn {
  readonly top: number;
  readonly bottom: number;
  readonly left: number;
  readonly right: number;
}

export interface PageSetup {
  readonly paperSize: "A3";
  readonly orientation: "landscape";
  readonly fitToPage: true;
  readonly fitToWidth: 1;
  readonly fitToHeight: 1;
  readonly marginsIn: PageMarginsIn;
  readonly printArea: RangeRef;
  /** Excel's own "zoom on open" (SPEC.md §3.3: 70%), independent of fitToPage's print scale. */
  readonly zoomPercent: number;
}

export interface SheetDescriptor {
  readonly name: string;
  readonly columns: readonly ColumnDef[];
  readonly rows: readonly RowDef[];
  readonly merges: readonly MergedRange[];
  readonly cells: readonly CellData[];
  readonly images: readonly ImagePlacement[];
  readonly pageSetup: PageSetup;
  readonly freezePanes: false;
  readonly gridlinesVisible: false;
}

/**
 * SPEC.md §2.3: a block's printed cell budget can be exceeded by the sum of
 * its `primary` entries. The descriptor records *what happened*, not what
 * to do about it — the "move oldest to appendix" fix is a UI/command
 * concern (Phase 3's command dispatcher), never something buildA3Layout
 * decides on its own.
 */
export interface OverflowWarning {
  /** Usually one step; [5, 6] for the template block that merges "develop" and "implement". */
  readonly stepIds: readonly StepId[];
  readonly budgetPt: number;
  readonly contentPt: number;
  readonly overflowByPt: number;
  /** Entry ids that did not fit inside the block's budget, in the order they were dropped. */
  readonly droppedEntryIds: readonly string[];
}

/**
 * G3 (D-198): SPEC.md §1.2 — "the A3 preview marks the step 'provisional'"
 * when its `readiness` evaluation is flagged. Mirrors `OverflowWarning`'s own
 * shape: the descriptor records *which block, what rectangle*, never *how to
 * draw it* — `HtmlA3Renderer` and the Rust writer each render their own
 * visual treatment (a dashed outline, D-198) from this geometry alone.
 */
export interface ProvisionalBlockMarker {
  /** Usually one step; e.g. [5, 6] for the template block that merges "develop" and "implement". */
  readonly stepIds: readonly StepId[];
  /** The block's full printed rectangle — header row through its last content row, e.g. "B7:O21". */
  readonly range: RangeRef;
}

/**
 * Faz 11/L3b (D-170): geometry for every `.elastic`-declared block in THIS
 * project's resolved layout — the drag-handle overlay (`BlockPinOverlay`,
 * outside `HtmlA3Renderer`/D-94, shared by `RightPanel`'s in-panel preview
 * and the pop-out `A3PreviewWindow`, D-133) needs a block's rectangle and
 * floor to draw a handle and a "reset to automatic"/at-floor indicator,
 * without recomputing anything `resolveElasticBlocks` already resolved. A
 * non-elastic block (every `farplas-7step-tr` block) never appears here.
 */
export interface ElasticBlockGeometry {
  /** Usually one step; every `pps-8step-auto` block is 1:1 with a single app-step (D-224). */
  readonly stepIds: readonly StepId[];
  readonly contentColumns: { readonly first: string; readonly last: string };
  readonly headerRange: RangeRef;
  readonly contentRows: { readonly start: number; readonly end: number };
  readonly minimumCanvasRows: number;
  /**
   * The raw value from `ProjectModel.blockPins` when this block currently
   * carries a manual override — absent otherwise. May differ from this
   * block's actual resolved `contentRows` span if the solver had to clamp
   * it down to protect a neighbour's own floor (D-170's iron law).
   */
  readonly pinnedCanvasRows?: number;
}

export interface A3LayoutDescriptor {
  readonly templateId: string;
  readonly language: "tr" | "en";
  readonly styles: readonly CellStyle[];
  readonly sheets: {
    readonly a3: SheetDescriptor;
    readonly appendices: readonly SheetDescriptor[];
  };
  readonly overflowWarnings: readonly OverflowWarning[];
  readonly provisionalBlocks: readonly ProvisionalBlockMarker[];
  readonly elasticBlocks: readonly ElasticBlockGeometry[];
}

File: src-tauri/src/xlsx/descriptor.rs
//! Serde mirror of `src/a3/descriptor.ts`. D-04: this module holds data
//! shapes only — no layout decisions. Every struct is `rename_all =
//! "camelCase"` so the JSON crossing the Tauri IPC boundary needs no
//! per-field renames to match the TypeScript source of truth.

use serde::Deserialize;

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CellFont {
    pub name: String,
    pub size_pt: f64,
    #[serde(default)]
    pub bold: bool,
    #[serde(default)]
    pub italic: bool,
    pub color: Option<String>,
}

#[derive(Debug, Clone, Copy, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum BorderWeight {
    Thin,
    Medium,
    Thick,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CellBorder {
    pub top: Option<BorderWeight>,
    pub bottom: Option<BorderWeight>,
    pub left: Option<BorderWeight>,
    pub right: Option<BorderWeight>,
    pub color: Option<String>,
}

#[derive(Debug, Clone, Copy, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum HorizontalAlign {
    Left,
    Center,
    Right,
}

#[derive(Debug, Clone, Copy, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum VerticalAlign {
    Top,
    Center,
    Bottom,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CellStyle {
    pub id: String,
    pub font: CellFont,
    pub fill_color: Option<String>,
    pub horizontal_align: Option<HorizontalAlign>,
    pub vertical_align: Option<VerticalAlign>,
    #[serde(default)]
    pub wrap_text: bool,
    pub border: Option<CellBorder>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(untagged)]
pub enum CellValue {
    Text(String),
    Number(f64),
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CellData {
    #[serde(rename = "ref")]
    pub cell_ref: String,
    pub value: Option<CellValue>,
    pub style_id: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct MergedRange {
    pub range: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ColumnDef {
    pub key: String,
    pub char_width: f64,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RowDef {
    pub index: u32,
    pub height_pt: f64,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImagePlacement {
    pub id: String,
    /// Base64-encoded image bytes. The writer never reads from disk (D-04).
    pub data: String,
    pub mime_type: String,
    pub anchor_cell: String,
    #[serde(default)]
    pub offset_x_pt: f64,
    #[serde(default)]
    pub offset_y_pt: f64,
    pub width_pt: f64,
    pub height_pt: f64,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PageMarginsIn {
    pub top: f64,
    pub bottom: f64,
    pub left: f64,
    pub right: f64,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PageSetup {
    pub paper_size: String,
    pub orientation: String,
    pub fit_to_page: bool,
    pub fit_to_width: u16,
    pub fit_to_height: u16,
    pub margins_in: PageMarginsIn,
    pub print_area: String,
    pub zoom_percent: u16,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SheetDescriptor {
    pub name: String,
    pub columns: Vec<ColumnDef>,
    pub rows: Vec<RowDef>,
    pub merges: Vec<MergedRange>,
    pub cells: Vec<CellData>,
    pub images: Vec<ImagePlacement>,
    pub page_setup: PageSetup,
    pub freeze_panes: bool,
    pub gridlines_visible: bool,
}

#[derive(Debug, Clone, Deserialize)]
pub struct Sheets {
    pub a3: SheetDescriptor,
    pub appendices: Vec<SheetDescriptor>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OverflowWarning {
    pub step_ids: Vec<u8>,
    pub budget_pt: f64,
    pub content_pt: f64,
    pub overflow_by_pt: f64,
    pub dropped_entry_ids: Vec<String>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct A3LayoutDescriptor {
    pub template_id: String,
    pub language: String,
    pub styles: Vec<CellStyle>,
    pub sheets: Sheets,
    pub overflow_warnings: Vec<OverflowWarning>,
}
